src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -217,23 +222,37 @@ protected Object invokeStep(StepDescriptor d, Object args) {
      * @param args The arguments passed to the step.
      */
     protected Object invokeStep(StepDescriptor d, String name, Object args) {
-        final NamedArgsAndClosure ps = parseArgs(args, d);
-
-        CpsThread thread = CpsThread.current();
-
-        FlowNode an;
 
         // TODO: generalize the notion of Step taking over the FlowNode creation.
         boolean hack = d instanceof ParallelStep.DescriptorImpl || d instanceof LoadStep.DescriptorImpl;
 
-        if (ps.body == null && !hack) {
-            an = new StepAtomNode(exec, d, thread.head.get());
+        FlowNode an;
+        CpsThread thread = CpsThread.current();
+        boolean hasBody = false;
+        if (!hack) {
+            if (args != null && args.getClass().isArray()) {
+                int size = Array.getLength(args);
+                if (size > 0 && Array.get(args, size - 1) instanceof CpsClosure) {
```

Comment:
nit: You could also check for `Object[]` directly to avoid the reflective APIs:

```suggestion
            if (args instanceof Object[]) {
                Object[] array = (Object[])args;
                if (array.length > 0 && array[array.length - 1] instanceof CpsClosure) {
```

Reply:
addressed in ade619f